### PR TITLE
Phpunit speedtrap

### PR DESCRIPTION
**Pull request by szepeviktor**
The purpose of this Pr is to measure long-running unit tests.

Was #12760

### CODE REVIEW

**Comment by szepeviktor** [comment]
# Results

### SystemTests

```
You should really fix these slow tests (>500ms)...
 1. 215204ms to run Piwik\Tests\System\ArchiveCronTest:testArchivePhpCron
 2. 59605ms to run Piwik\Tests\System\TwoVisitorsTwoWebsitesDifferentDaysTest:testApi with data set #7
 3. 54062ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #92
 4. 51416ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #96
 5. 51372ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #90
 6. 50215ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #74
 7. 48881ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #72
 8. 48671ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #94
 9. 47777ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #66
 10. 47513ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #58
...and there are 202 more above your threshold hidden from view
Time: 32.84 minutes, Memory: 368.00MB
```

**Comment by szepeviktor** [comment]
Something very fundamental takes ~50 per unit test

**Comment by mattab** [comment]
Thank you @szepeviktor for creating your first PR, that's really useful :+1: making our tests faster has become top priority as the System tests often fail as they exceed 40min. 

From this output it looks like possibly `Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi` should be tackled first as this causes probably 15+min or more in itself. (these tests check that each dimension can be used as segment, and have at least a visit matching in the fixture.)

```
You should really fix these slow tests (>500ms)...
 1. 215204ms to run Piwik\Tests\System\ArchiveCronTest:testArchivePhpCron
 2. 59605ms to run Piwik\Tests\System\TwoVisitorsTwoWebsitesDifferentDaysTest:testApi with data set #7
 3. 54062ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #92
 4. 51416ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #96
 5. 51372ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #90
 6. 50215ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #74
 7. 48881ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #72
 8. 48671ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #94
 9. 47777ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #66
 10. 47513ms to run Piwik\Tests\System\AutoSuggestAPITest:testAnotherApi with data set #58
```

**Comment by szepeviktor** [comment]
@mattab Do you have a 1 page summary on how to get from cloning matomo repo to running a test?
Maybe a Dockerfile??

**Comment by sgiehl** [comment]
this might help: https://github.com/matomo-org/matomo/blob/3.x-dev/tests/README.md

**Comment by szepeviktor** [comment]
I think I need to create a Dockerfile to be able to *breath*

**Comment by szepeviktor** [comment]
Could you help me how `piwik_tests` gets populated on Travis?

**Comment by szepeviktor** [comment]
Adding comments to the testing system (travis.yml, shell scripts) may help a lot.

**Comment by diosmosis** [comment]
@szepeviktor I use docker half of the time for development: https://github.com/diosmosis/matomo-docker-dev . It's definitely not official, so it may or may not work for you, I can't promise any real support and I will make changes to it as I need them.

Use `docker-compose up` to start it up (and all the other containers I use when developing). Install matomo by visiting `localhost:3000`, and run commands via the `console` script in the readme, eg, `/path/to/matomo-docker-dev/console tests:run ./tests/PHPUnit/System/OneVisitorTwoVisitsTest.php` (this will run the `console` php script in the matomo container).

As for how the test database is populated, it is created, setup & filled by the https://github.com/matomo-org/matomo/blob/3.x-dev/tests/PHPUnit/Framework/Fixture.php class and its descendants. (Fair warning, it sets up the entire matomo test environment, and so is a very large, hard to read class, and may make your eyes bleed.) Some test setup code is located elsewhere like in the https://github.com/matomo-org/matomo/blob/3.x-dev/tests/PHPUnit/Framework/TestingEnvironmentManipulator.php class.

**Comment by szepeviktor** [comment]
Thank you.

File: .travis.yml [hunk 1]
```diff
@@ -46,38 +46,11 @@ matrix:
...
-    - php: 5.6
-      env: TEST_SUITE=AllTests MYSQL_ADAPTER=MYSQLI ALLTEST_EXTRA_OPTIONS="--run-second-half-only"
-      sudo: required
-    # UITests use a specific version because the default 5.5 (== 5.5.38) is missing FreeType support
-    - php: 5.5.33
-      env: TEST_SUITE=UITests MYSQL_ADAPTER=PDO_MYSQL UITEST_EXTRA_OPTIONS="--run-first-half-only"
-      sudo: false
-    - php: 5.5.33
-      env: TEST_SUITE=UITests MYSQL_ADAPTER=PDO_MYSQL UITEST_EXTRA_OPTIONS="--run-second-half-only"
-      sudo: false
```
Comment (diosmosis) [inline .travis.yml]: These changes need to be reverted before this PR can be merged.

**Review by diosmosis — CHANGES_REQUESTED** [review]
* The minimum execution time threshold needs to be increased for system tests, as they will always be slower than unit tests. Not sure what the best value is since some will always be long running. Perhaps it would be good to make the speed trap listener optional in some way.
* There's a conflict in the composer file that needs to be fixed.

**Comment by szepeviktor** [comment]
Thank you @diosmosis!

This PR is a rather like a debugging session.
It's goal is to find out what takes ~50 second per test.
Up to now I wasn't able to find it out.

**Comment by diosmosis** [comment]
@szepeviktor Gotcha, will remove this from the current milestone then, feel free to create PRs when you find some optimizations :)

**Comment by mattab** [comment]
Thanks @szepeviktor for creating this PR. I've added a comment in the related issue here: https://github.com/matomo-org/matomo/issues/12691#issuecomment-400475879
I'll close this PR now, but feel free to re-open it as we could in theory add the speedtrap by default in Matomo.

**Pull request closed without merging**